### PR TITLE
Toggle bucket logging

### PR DESCRIPTION
**Pull request by jdubs**
Toggle remote s3 access logging to a remote bucket. Remote bucket ACLs must permit read and Write for log delivery. Not specifying a target_prefix will default to the current bucket in question.

policies:
  - name: s3-enable-logging
    resource: s3
    filter:
      - "tag:Testing": present
    actions:
      - type: toggle-logging
        target_bucket: log-bucket
        target_prefix: logs123

### CODE REVIEW

File: c7n/resources/s3.py [hunk 1]
```diff
@@ -502,6 +502,57 @@ def process(self, resources):
...
+        target_bucket={'type': 'string'},
+        target_prefix={'type': 'string'},
+    )
+
+    def process(self, resources):
+        enabled = self.data.get('enabled', True)
+        client = local_session(self.manager.session_factory).client('s3')
+        for r in resources:
+            target_prefix = self.data.get('target_prefix', r['Name'])
+            if 'TargetPrefix' in r['Logging']:
```
Comment (kapilt) [inline c7n/resources/s3.py]: target prefix isn't required for logging to be active, targetbucket is.

**Comment by jdubs** [comment]
Good catch. I'll fix that next year!

**Comment by jdubs** [comment]
A quick question for you. I'm trying to do something like the following to set a bucket logging endpoint for buckets that already do not have logging enabled. The s3 api sends back a blank dictionary when logging is not enabled and a dictionary when it's enabled like so.

```
aws s3api get-bucket-logging --bucket  jtrenholm-bucket-testing
{
    "LoggingEnabled": {
        "TargetPrefix": "jtrenholm-bucket-testing", 
        "TargetBucket": "super-log"
    }
}
```
From my understanding I should be able to trigger the bucket logging action when `LoggingEnabled` is absent. However the filter doesn't catch it and I do not understand why.
Thoughts?

Example policy. I've also tried numerous different value filters ranging from, TargetBucket not in LoggingEnabled etc and has not had any effect.
```
policies:
  - name: s3-enable-logging
    description: |
        Enable s3 access logging for all buckets with out logging enabled. Ship to the target bucket. Access logs are created in target-bucket/<current-s3-bucket>
    resource: s3
    filters:
      - type: value
        key: LoggingEnabled
        value: absent
    actions:
      - type: toggle-logging
        target_bucket: *logging_bucket
```

**Comment by tim-elliott** [comment]
What the aws cli returns isn't always what custodian returns.  In this case the custodian output looks like 
`    "Logging": {
      "TargetPrefix": "logs/", 
      "TargetBucket": "some-bucket"
    },`

So one way to check that there is no logging is:
`    filters:
      - type: value
        key: "Logging.TargetBucket"
        value: absent`
You can't check if Logging alone is absent, because even if there is no content
`    "Logging": {},`
the key "Logging" is always there.

**Comment by jdubs** [comment]
Thank you for clearing that up for me.

**Comment by kapilt** [comment]
when custodian assembles the information for an s3 bucket it put the s3 bucket subdocuments values into an arbitrary set of keys...

The mapping of subdocuments to keys https://github.com/capitalone/cloud-custodian/blob/master/c7n/resources/s3.py#L110 

These should be added to the s3 resource docs. its unfortunate that we didn't just use the subdocument keys names. at some point i'd like to get to validating key expressions against the resource schema.

using a filter with value `not-null` will still match those without logging.

**Pull request merged**